src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs

```diff
@@ -74,8 +76,250 @@ public static async ValueTask StoreAsync(string key, OutputCacheEntry value, Tim
 
         using var bufferStream = new MemoryStream();
 
-        JsonSerializer.Serialize(bufferStream, formatterEntry, FormatterEntrySerializerContext.Default.FormatterEntry);
+        Serialize(bufferStream, formatterEntry);
 
         await store.SetAsync(key, bufferStream.ToArray(), value.Tags ?? Array.Empty<string>(), duration, cancellationToken);
     }
+
+    // Format:
+    // Serialization revision:
+    //   7-bit encoded int
+    // Creation date:
+    //   Ticks: 7-bit encoded long
+    //   Offset.TotalMinutes: 7-bit encoded long
+    // Status code:
+    //   7-bit encoded int
+    // Headers:
+    //   Headers count: 7-bit encoded int
+    //   For each header:
+    //     key name byte length: 7-bit encoded int
+    //     UTF-8 encoded key name byte[]
+    //     Values count: 7-bit encoded int
+    //     For each header value:
+    //       data byte length: 7-bit encoded int
+    //       UTF-8 encoded byte[]
+    // Body:
+    //   Segments count: 7-bit encoded int
+    //   For each segment:
+    //     data byte length: 7-bit encoded int
+    //     data byte[]
+    // Tags:
+    //   Tags count: 7-bit encoded int
+    //   For each tag:
+    //     data byte length: 7-bit encoded int
+    //     UTF-8 encoded byte[]
+
+    private static void Serialize(Stream output, FormatterEntry entry)
+    {
+        using var writer = new BinaryWriter(output);
```

Comment:
There's allocations for the MemoryStream and the BinaryWriter. Should these be cached / pooled?

(similar for Deserialize)

Reply:
I don't think these should be pooled, they don't have reusable state. Ideally we should use a custom `MemoryStream` that pools its buffers (there is an issue for that in the runtime). Or change the API design such that we can use streams with the cache store instead of returning `byte[]`.